Source/WebCore/page/ElementTargetingController.cpp

```diff
@@ -978,58 +996,95 @@ void ElementTargetingController::applyVisibilityAdjustmentFromSelectors(Document
     auto viewportArea = m_viewportSizeForVisibilityAdjustment.area();
     Region adjustmentRegion;
     Vector<String> matchingSelectors;
-    for (auto& selectorIncludingPseudo : *m_remainingVisibilityAdjustmentSelectors) {
-        auto [selector, adjustment] = resolveSelectorToQuery(selectorIncludingPseudo);
-        if (selector.isEmpty()) {
-            // FIXME: Handle the case where the full selector is `::after|before`.
-            continue;
-        }
-
-        auto queryResult = document.querySelector(selector);
-        if (queryResult.hasException())
-            continue;
-
-        RefPtr element = queryResult.releaseReturnValue();
-        if (!element)
+    for (auto& selectorsForElementIncludingShadowHosts : *m_remainingVisibilityAdjustmentSelectors) {
+        if (selectorsForElementIncludingShadowHosts.isEmpty())
             continue;
 
-        CheckedPtr renderer = element->renderer();
-        if (!renderer)
-            continue;
+        bool foundLastTarget = false;
+        Ref<ContainerNode> containerToQuery = document;
+        for (size_t index = 0; index < selectorsForElementIncludingShadowHosts.size(); ++index) {
```

Comment:
It seems like we could use an iterator here rather than an index based loop?

Reply:
Ah, yeah — I only used an index here to know when I've resolved the final target, but I can track that index separately, and use a iterator.